app_dart/lib/src/service/cache_service.dart

```diff
@@ -20,6 +21,8 @@ class CacheService {
   }) : _provider =
             inMemory ? Cache.inMemoryCacheProvider(inMemoryMaxNumberEntries) : Cache.redisCacheProvider(memorystoreUri);
 
+  final ReadWriteMutex m = ReadWriteMutex();
```

Comment:
1. package:mutex will only work locally. During peak hours, Cocoon can have 5 instances. I think this is worth landing as it's simpler than implementing a datastore lock, but keep in mind we may need to continue working on this.
2. Is this lock universal to all cache entries? If so, there is bound to be a lot of contention in it's current state

Reply:
Hmmm, I see what you are saying. 

This guards against the multiple calls locally but yeah we will still have possible issues when we are attempting to store the new tokens in the app installation access tokens storage.  

We should create a token service.

Yes the lock blocks the cache for all entries, not just tokens. Perhaps we need to keep multiple caches for tokens and other values to keep them separate.

Reply:
Opening an issue for further work. I will link it here.